### PR TITLE
Replaced BoxedSystem param with SystemRefMut

### PR DESCRIPTION
**Pull request by MNTRA**
Removes the requirement that systems must be boxed when calling `SystemStageExecutor::execute_stage(..)`.
This gives greater flexibility to Stages on how they wish to store their systems as they can now dynamically create a slice of `SystemRefMut` instead of being forced to store a `Vec<BoxedSystem>` or moving the system into a slice before calling `SystemStageExecutor::execute_stage(..)`. This change also means that stages can now dynamically choose which systems to run by storing a bank of systems and composing a slice of references they wish to pass to the executor.

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/stage.rs [hunk 1]
```diff
@@ -107,9 +107,14 @@ impl SystemStage {
     }
 
     pub fn run_once(&mut self, world: &mut World, resources: &mut Resources) {
+        let mut system_refs = Vec::with_capacity(self.systems.len());
+        for system in self.systems.iter_mut() {
+            system_refs.push(system.as_mut());
+        }
```
Comment (bjorn3) [inline crates/bevy_ecs/src/schedule/stage.rs:113]: This would allocate a new vec every time the stage runs, right? Maybe pass an iterator instead?

**Comment by alice-i-cecile** [comment]
This proposal should probably wait until #1144 is merged to be implemented. Pinging @Ratysz for thoughts on the implications of this approach.

**Comment by Ratysz** [comment]
> [...] thoughts on the implications of this approach.

On its own, I'm fairly sure it doesn't solve anything, yet introduces the extra allocations @bjorn3 points out. I assume this makes more sense with a bespoke stage reimplementation, one that doesn't store systems as trait objects, but given that the entire notion of stage executor is completely reworked in aforementioned #1144 this is not needed.

**Comment by cart** [comment]
Yeah I do think this increases the flexibility of the current api, which is nice (if we can sort out a way to remove the per-update allocations). But as @Ratysz said, given that this api is reworked in #1144, we should close this.

@MNTRA I do like where your head is that though. Keep looking for stuff like this!

**Pull request closed without merging**